`World.spawn()`, Non-primitive Assignments, and Patching

zicklag opened the pull request:
This PR

1. Adds a simple `world.spawn()` method.
1. Allows you to assign non-primitives to value refs, allowing, in my use case, me to set the `id` of a `Handle<T>` in Bevy even though `HandleId` is not a JS primitive.
2. Fixes a bug I introduced with the `Value.patch()` PR where non-primitive assignments stopped throwing an error when they failed.

It's worth noting that non-primitive patching doesn't work yet. That will require a kind of recursive value ref unwrapping helper, which I haven't implemented yet.

zicklag committed:
Add world.spawn()

zicklag committed:
Allow Assigning Non-Primitive Value Refs

zicklag committed:
Implement Non-primitive Patching

zicklag committed:
Strictly Type patch() and create() Methods

zicklag commented:
Just pushed an update that adds non-primitive patching.

This let me load ( weak ) asset handles in my game quite ergonomically:

```ts
type AnimatedSprite = {
  start: usize;
  end: usize;
  atlas: HandleTextureAtlas;
  flip_x: boolean;
  flip_y: boolean;
  repeat: boolean;
  fps: f32;
};
const AnimatedSprite: BevyType<AnimatedSprite> = {
  typeName: "jumpy::animation::AnimatedSprite",
};

export default {
  preUpdate() {
    for (const entity of MapElement.getSpawnedEntities()) {
      let animated_sprite = Value.create(AnimatedSprite, {
        start: 0,
        end: 4,
        repeat: true,
        fps: 6,
        atlas: {
          id: Assets.getHandleId(
            "../../../resources/default_decoration.atlas.yaml"
          ),
        },
      });

      world.insert(entity, animated_sprite);
    }
  },
};

```

zicklag commented:
If https://github.com/bevyengine/bevy/pull/6187 is merged, I think we'll be able to spawn sprites from scripts, with the nuance that the script can only create weak handles, so the game needs to provide a way for scripts to somehow register assets to load with strong handles.

In my game I have YAML files with a `preload_assets` field for that, because I preload all the assets in my game anyway.

jakobhellermann commented:
Have you seen https://github.com/bevyengine/bevy/pull/5923? It would let you insert new assets and get back a handle.

zicklag commented:
No I haven't, that's awesome! 🎉

I think that pretty much fixes the bulk of the things I was worried about for asset handling in scripts.

jakobhellermann commented on src/runtime/ops/ecs/value.rs:
```diff
@@ -244,22 +362,53 @@ pub fn ecs_value_ref_set(
+            let new_value_ref = value_refs
+                .get(new_js_value_ref.key)
+                .ok_or_else(|| format_err!("Value ref doesn't exist"))?;
+            let new_reflect = new_value_ref.get(world)?.clone_value();
```
Okay so this implicitly clones the value, right? I'm not sure I like that.
I would expect
```rust
let val = transform.translation;
val.x = 5;
info(val.translation.x)
```
to print 5, and any clones be explicit, e.g. like `Value.clone(val.translation)`.

jakobhellermann replied:
This would of course benefit from `ReflectClone` instead of `clone_dynamic` which outputs the `DynamicX` proxy types.

zicklag replied:
That does print 5 just like you are expecting.

The clone is actually simply to avoid borrowing issues, because we aren't allowed to take an immutable reference to `world` when running `new_value_ref.get(world)` and then hold that ref while we mutably borrow world with `value_ref.get_mut(world)`.

It doesn't change semantics, but it does obviously result in an extra clone as far as performance is concerned.

I was wondering whether or not there would be a way to somehow safely borrow multiple value refs from the world at the same time, having it make sure the borrows don't alias when you do the `borrow_multiple()` or something like that.

---

And yeah, a `ReflectClone` would be great. :)

zicklag committed:
Clear Value Refs Every Frame

zicklag committed:
Re-export bevy_reflect_fns

zicklag committed:
Try Downcasting Function Calls to Primitive Values

zicklag committed:
Set Script Info During Script Load on WASM

zicklag committed:
Fix Integer Assignments on Native

When trying to assign integers on native, there would be a serialization
error because the numbers were deserialized with an extra decimal.

This converts valid integers to integers during deserialization of the
arguments to make sure integer assignments succeed.

zicklag committed:
Avoid Error About Not Reading Prop `bind` of null

jakobhellermann commented on types/lib.bevy.d.ts:
```diff
@@ -22,9 +22,13 @@ declare interface BevyScript {
 declare type RawValueRef = unknown;
 
+type RecursivePartial<T> = {
+    [P in keyof T]?: RecursivePartial<T[P]>;
```
doesn't this need some kind of recursion base condition? `RecursivePartial<{x: number, y: number, z: number}>` would become `{x: RecursivePartial<number>, y: RecursivePartial<number>, z: RecursivePartial<number> }`, at which point you get `keyof number` which apparently is `"toString" | "toFixed" | ..`

zicklag replied:
We could do this:

```ts
type RecursivePartial<T> = {
  [P in keyof T]?: T extends object ? RecursivePartial<T[P]> : T[P];
};
```

But interestingly, when I hover over the type inference in VSCode, it seems like it's correct, even without the `T extends object` check.

For instance, without changing it, this doesn't work:

```ts
const a: RecursivePartial<number> = { toString: (test) => ""};
```

And when I hover over `q` in this example in VSCode, it shows the type of `name` to be `string`, not `RecursivePartial<string>`:

```ts
const test: RecursivePartial<{
  x: number;
  y: number;
  z: number;
  q: {
    name: string;
  };
}> = {
  x: 3,
  q: {

  }
};
```

![image](https://user-images.githubusercontent.com/25393315/198836467-30d0d929-40f3-4620-9dbd-99d55d1e7096.png)

And this won't work:

```ts
const test: RecursivePartial<{
  x: number;
  y: number;
  z: number;
  q: {
    name: string;
  };
}> = {
  x: 3,
  q: {
    toString: (_) => {""}
  }
};
```

So it seems like TypeScript is doing some kind of automatic transformation on the type or something, making it do what we'd expect.

jakobhellermann commented:
Can you rebase this? Otherwise it looks good. Sorry for leaving these PRs uncommented for so long.

zicklag commented:
Yep, I'll rebase this today, and no problem about the delay. I was totally fine working off my fork.

zicklag commented:
Pushed the rebase.

jakobhellermann commented:
Thanks a ton for the PR and the patience for having this unmerged for so long.
I rebased the PR on main and merged it in https://github.com/jakobhellermann/bevy_mod_js_scripting/pull/36.

zicklag commented:
No problem! :D
